sympy/solvers/solvers.py

```diff
@@ -863,64 +857,141 @@ def _sympified_list(w):
                         symbols.add(p)
                         pot.skip()  # don't go any deeper
         symbols = list(symbols)
-
         ordered_symbols = False
-    elif len(symbols) == 1 and iterable(symbols[0]):
-        symbols = symbols[0]
+    else:
+        # User has given symbols as *args or as an iterable arg
+        if len(symbols) == 1 and iterable(symbols[0]):
+            symbols = symbols[0]
+            if isinstance(symbols, GeneratorType):
+                symbols = list(symbols)
+        # Has the user given an unordered collection like a set?
+        ordered_symbols = is_sequence(symbols)
+        symbols = [sympify(sym) for sym in symbols]
+
+    # Remove symbols the user is not interested in. This removes symbols even
+    # if they were explicitly specified.
 
-    # remove symbols the user is not interested in
-    exclude = flags.pop('exclude', set())
     if exclude:
-        if isinstance(exclude, Expr):
+        if not iterable(exclude):
             exclude = [exclude]
         exclude = set().union(*[e.free_symbols for e in sympify(exclude)])
-    symbols = [s for s in symbols if s not in exclude]
+        symbols = [s for s in symbols if s not in exclude]
 
+    # alternate exits
 
-    # preprocess equation(s)
-    ###########################################################################
-    for i, fi in enumerate(f):
-        if isinstance(fi, (Equality, Unequality)):
-            if 'ImmutableDenseMatrix' in [type(a).__name__ for a in fi.args]:
-                fi = fi.lhs - fi.rhs
-            else:
-                L, R = fi.args
-                if isinstance(R, BooleanAtom):
-                    L, R = R, L
-                if isinstance(L, BooleanAtom):
-                    if isinstance(fi, Unequality):
-                        L = ~L
-                    if R.is_Relational:
-                        fi = ~R if L is S.false else R
-                    elif R.is_Symbol:
-                        return L
-                    elif R.is_Boolean and (~R).is_Symbol:
-                        return ~L
-                    else:
-                        raise NotImplementedError(filldedent('''
-                            Unanticipated argument of Eq when other arg
-                            is True or False.
-                        '''))
+    def is_eq_bool(eq):
+        if isinstance(eq, (Equality, Unequality)):
+            return any(side in (S.false, S.true) for side in eq.args)
+
+    if any(is_eq_bool(fi) for fi in f):
```

Comment:
Can we move everything in this if block out to a separate function?

I think this "feature" of solve should probably be removed. Let's at least arrange the code so that it is clear what it is doing and how it is separate from everything else.